syntax = "proto3";

package grpc_service;

service GrpcService {
  rpc Ping (PingRequest) returns (PingResponse) {}
  rpc ServerStreamData (DataRequest) returns (stream DataChunk) {}
  rpc ClientStreamData (stream DataChunk) returns (TransferStatus) {}
  rpc NfsGetAttr (NfsGetAttrRequest) returns (NfsGetAttrResponse) {} // Added NfsGetAttr RPC
  rpc NfsReadDir (NfsReadDirRequest) returns (NfsReadDirResponse) {} // Added NfsReadDir RPC
}

message PingRequest {
  string message = 1;
}

message PingResponse {
  string message = 1;
}

message DataChunk {
  bytes data = 1;
}

message TransferStatus {
  bool success = 1;
  int64 bytes_received = 2;
}

message DataRequest {
  int64 size = 1;
}

//======================================================================
// New messages for NfsGetAttr
message NfsGetAttrRequest {
  string path = 1; // Path for which attributes are requested
}

message NfsGetAttrResponse {
  bool success = 1; // Indicates if the operation was successful
  string message = 2; // Message for additional information
  int64 size = 3; // Size of the file
  int32 mode = 4; // File mode (permissions)
  int32 nlink = 5; // Number of hard links
}

//======================================================================
// New messages for NfsReadDir
message NfsReadDirRequest {
  string path = 1; // Path of the directory to read
}

message NfsReadDirResponse {
  bool success = 1; // Indicates if the operation was successful
  string message = 2; // Message for additional information
  repeated string files = 3; // List of files in the directory
}